components/files/constrainedfilestream.cpp

```diff
@@ -103,21 +103,23 @@ namespace Files
 
     };
 
-    ConstrainedFileStream::ConstrainedFileStream(const char *filename, size_t start, size_t length)
-        : std::istream(new ConstrainedFileStreamBuf(filename, start, length))
+    ConstrainedFileStream::ConstrainedFileStream(std::streambuf* buf)
+        : std::istream(buf)
+        , mBuf(buf)
     {
 
     }
 
     ConstrainedFileStream::~ConstrainedFileStream()
     {
-        delete rdbuf();
+        delete mBuf;
     }
 
 
     IStreamPtr openConstrainedFileStream(const char *filename,
                                                        size_t start, size_t length)
     {
-        return IStreamPtr(new ConstrainedFileStream(filename, start, length));
+        ConstrainedFileStreamBuf* buf = new ConstrainedFileStreamBuf(filename, start, length);
```

Comment:
This looks mildly unsafe. If the second new happens to throw, the first one would be left without a matching delete, i.e. resources leak.

Reply:
What can we do here?

Reply:
I'd recommend a `unique_ptr`, unless you can't get away with multiple owners.

Reply:
Actually, there are two owners for now: mBuf field and openConstrainedFileStream() function.
I tried to use the unique_ptr here, but got the crash.

Reply:
`openConstrainedFileStream` just passes it to the `ConstrainedFileStream` constructor and relinquishes ownership. Did you use `std::move` to pass it?

Reply:
> Did you use std::move to pass it?

No, I did not. It seems it led to crash. With std::move it works.